File: Fbs.WebApi/Repository/NominalRollRepository.cs
using System.Linq.Expressions;
using Fbs.WebApi.Entities;
using Fbs.WebApi.Options;
using Google.Apis.Sheets.v4;
using Microsoft.Extensions.Caching.Hybrid;
using Microsoft.Extensions.Options;

namespace Fbs.WebApi.Repository;

public class NominalRollRepository(
    InstrumentationSource instrumentation,
    HybridCache cache,
    IOptions<GoogleOptions> options,
    SheetsService sheetsService
) : IRepository<NominalRoll>
{
    private readonly string[] _header =
    [
        "Name",
        "Unit",
        "Phone",
    ];

    public async Task<List<NominalRoll>> GetListAsync(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await cache.GetOrCreateAsync(
            "Nominal Roll",
            (sheetsService, options),
            async (state, ct) =>
                await state.sheetsService.Spreadsheets.Values.Get(state.options.Value.SpreadsheetId, "Nominal Roll")
                    .ExecuteAsync(ct),
            cancellationToken: cancellationToken);

        if (!items.Values.First().SequenceEqual(_header))
        {
            throw new Exception("Unexpected headers in Nominal Roll table");
        }

        return items.Values
            .Skip(1)
            .Select((row, idx) => new NominalRoll 
            {
                Row = idx + 2,
                Name = row.ElementAtOrDefault(0) as string,
                Unit = row.ElementAtOrDefault(1) as string,
                Phone = row.ElementAtOrDefault(2) as string,
            })
            .ToList();
    }

    public async Task<NominalRoll?> FindAsync(Expression<Func<NominalRoll, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.SingleOrDefault(predicate.Compile());
    }

    public async Task<NominalRoll> GetAsync(Expression<Func<NominalRoll, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.Single(predicate.Compile());
    }

    public async Task<NominalRoll> InsertAsync(NominalRoll entity, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    public async Task<NominalRoll> UpdateAsync(NominalRoll entity, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    public async Task DeleteAsync(Expression<Func<NominalRoll, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }
}

File: Fbs.WebApi/src/main/kotlin/sg/from/fbs/repository/NominalRollRepository.kt
package sg.from.fbs.repository

import com.google.api.services.sheets.v4.Sheets
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository
import sg.from.fbs.config.GoogleProperties
import sg.from.fbs.entities.NominalRoll
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy

@Repository
class NominalRollRepository(
    private val sheetsService: Sheets,
    private val options: GoogleProperties
) {
    @Autowired
    @Lazy
    lateinit var self: NominalRollRepository

    private val header = listOf("Name", "Unit", "Phone")

    @Cacheable("Nominal Roll")
    fun getList(): List<NominalRoll> {
        val response = sheetsService.spreadsheets().values()
            .get(options.spreadsheetId, "Nominal Roll")
            .execute()

        val values = response.getValues()
        if (values == null || values.isEmpty()) return emptyList()

        val actualHeader = values.first().map { it.toString() }
        if (actualHeader != header) {
            throw Exception("Unexpected headers in Nominal Roll table")
        }

        return values.drop(1).mapIndexed { idx, row ->
            NominalRoll(
                row = idx + 2,
                name = row.getOrNull(0) as? String,
                unit = row.getOrNull(1) as? String,
                phone = row.getOrNull(2) as? String
            )
        }
    }

    fun find(predicate: (NominalRoll) -> Boolean): NominalRoll? {
        return self.getList().find(predicate)
    }

    fun get(predicate: (NominalRoll) -> Boolean): NominalRoll {
        return self.getList().first(predicate)
    }

    fun insert(entity: NominalRoll): NominalRoll {
        throw NotImplementedError()
    }

    fun update(entity: NominalRoll): NominalRoll {
        throw NotImplementedError()
    }

    fun delete(predicate: (NominalRoll) -> Boolean) {
        throw NotImplementedError()
    }
}